src/Symfony/Component/TypeInfo/TypeResolver/StringTypeResolver.php

```diff
@@ -0,0 +1,248 @@
+                return $type;
+            }
+
+            $variableTypes = array_map(fn (TypeNode $t): Type => $this->getTypeFromNode($t, $typeContext), $node->genericTypes);
+
+            if ($type instanceof CollectionType) {
+                $keyType = $type->getCollectionKeyType();
+
+                $type = $type->getType();
+                if ($type instanceof GenericType) {
+                    $type = $type->getType();
+                }
+
+                if (CollectionType::LIST_TYPE_COUNT === \count($variableTypes)) {
+                    return Type::collection($type, $variableTypes[0], $keyType);
+                } elseif (CollectionType::MAP_TYPE_COUNT === \count($variableTypes)) {
+                    return Type::collection($type, $variableTypes[1], $variableTypes[0]);
+                }
+            }
+
+            if ($type instanceof ObjectType && \in_array($type->getClassName(), [\Traversable::class, \Iterator::class, \IteratorAggregate::class, \ArrayAccess::class, \Generator::class], true)) {
+                return match (\count($variableTypes)) {
+                    CollectionType::LIST_TYPE_COUNT => Type::collection($type, $variableTypes[0]),
+                    CollectionType::MAP_TYPE_COUNT => Type::collection($type, $variableTypes[1], $variableTypes[0]),
+                    default => Type::collection($type),
+                };
+            }
+
+            return Type::generic($type, ...$variableTypes);
+        }
+
+        if ($node instanceof UnionTypeNode) {
+            return Type::union(...array_map(fn (TypeNode $t): Type => $this->getTypeFromNode($t, $typeContext), $node->types));
+        }
+
+        if ($node instanceof IntersectionTypeNode) {
+            return Type::intersection(...array_map(fn (TypeNode $t): Type => $this->getTypeFromNode($t, $typeContext), $node->types));
+        }
+
+        throw new \DomainException(sprintf('Unhandled "%s" node.', $node::class));
+    }
+
+    private function resolveCustomIdentifier(string $identifier, ?TypeContext $typeContext): Type
+    {
+        $classNameOrTemplate = $typeContext ? $typeContext->normalize($identifier) : $identifier;
+
+        if (isset(self::$classExistCache[$classNameOrTemplate])) {
+            return self::$classExistCache[$classNameOrTemplate] ? Type::object($classNameOrTemplate) : Type::template($classNameOrTemplate);
+        }
+
+        if (class_exists($classNameOrTemplate) || interface_exists($classNameOrTemplate)) {
+            self::$classExistCache[$classNameOrTemplate] = true;
+
+            return Type::object($classNameOrTemplate);
+        }
+
+        try {
+            new \ReflectionClass($classNameOrTemplate);
+            self::$classExistCache[$classNameOrTemplate] = true;
+
+            return Type::object($classNameOrTemplate);
+        } catch (\Throwable) {
+        }
+
+        self::$classExistCache[$classNameOrTemplate] = false;
+
+        return Type::template($classNameOrTemplate);
```

Comment:
For me personally, this is a bit unexpected to fallback to `TemplateType` for every identifier that isn't a known class and is not a pseudo-type already supported by the component.

Let's assume some static analysis / documentation tool introduces a new pseudo-type, e.g. `odd-number` and someone uses it in their code. Until support for this type is added to the component, the type for the node described with this new pseudo-type is resolved to `TemplateType` which is not true. Similar case is when someone makes a typo when typing a class name, e.g. `@param stdClas $foo` - it will be reported as `TemplateType`.

I don't know if there's a way to detect if the identifier is a pseudo-class or not, but maybe we shouldn't assume it's a template if we don't know what it really is.

See related issue #44455

Reply:
We should make a map with all pseudo-types and map them to a native type.
I would say this could be a nice first PR once this is merged since current component has the same features we currently have in Symfony.

Reply:
Actually, I think doing the following:

- First gather as much templates as possible with the TypeContextFactory (class templates are actually gathered but I'll gather method templates as well).
- Then, in the StringTypeResolver, if we find a template in the TypeContext, we return a TemplateType, and if not, we throw an exception.

WDYT?

Reply:
@Korbeil 
It won't solve the issue I described because it's impossible to know **all** pseudo-types that exist. The ones that are already resolved correctly are used by the most popular tools like phpstan, but there may be some less popular libraries that we don't know of. Nevertheless having such a map would be nice from code readability perspective probably but it's not crucial.

@mtarld 
That sounds much better in my opinion. I think throwing an exception is a safest result in this case.

Reply:
Here you go: https://github.com/symfony/symfony/pull/52510/files#diff-82935a7583645f83e164c76862126f68307e568b869f677980270a317e37f120R247-R251 :slightly_smiling_face: 

Thanks for that nice catch!